cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/deploy/InMemoryProgramRunDispatcher.java

```diff
@@ -0,0 +1,457 @@
+      File tempDir, @Nullable ApplicationSpecification appSpec) throws Exception {
+    // appSpec is null in an unit test
+    if (appSpec == null) {
+      return options;
+    }
+
+    Set<String> files = Sets.newHashSet();
+    ImmutableMap.Builder<String, String> builder = ImmutableMap.builder();
+    builder.putAll(options.getArguments().asMap());
+    for (Map.Entry<String, Plugin> pluginEntry : appSpec.getPlugins().entrySet()) {
+      Plugin plugin = pluginEntry.getValue();
+      File destFile = new File(tempDir, Artifacts.getFileName(plugin.getArtifactId()));
+      // Skip if the file has already been copied.
+      if (!files.add(destFile.getName())) {
+        continue;
+      }
+
+      try {
+        ArtifactId artifactId = Artifacts.toProtoArtifactId(programId.getNamespaceId(),
+            plugin.getArtifactId());
+        ArtifactDetail artifactDetail = getArtifactDetail(artifactId);
+        if (isDistributed) {
+          copyArtifactInDistributed(artifactId, artifactDetail, destFile);
+        } else {
+          copyArtifact(artifactDetail, destFile);
+        }
+      } catch (ArtifactNotFoundException e) {
+        throw new IllegalArgumentException(
+            String.format("Artifact %s could not be found", plugin.getArtifactId()), e);
+      }
+    }
+    LOG.debug("Plugin artifacts of {} copied to {}", programId, tempDir.getAbsolutePath());
+    builder.put(ProgramOptionConstants.PLUGIN_DIR, tempDir.getAbsolutePath());
+    return new SimpleProgramOptions(options.getProgramId(), new BasicArguments(builder.build()),
+        options.getUserArguments(), options.isDebug());
+  }
+
+  /**
+   * Copies the artifact jar to the given target file.
+   *
+   * @param artifactDetail detail information of the artifact to be copied
+   * @param targetFile target file to copy to
+   * @throws IOException if the copying failed
+   */
+  private void copyArtifact(ArtifactDetail artifactDetail, File targetFile) throws IOException {
+    Locations.linkOrCopy(artifactDetail.getDescriptor().getLocation(), targetFile);
+  }
+
+  private void copyArtifactInDistributed(ArtifactId artifactId, final ArtifactDetail artifactDetail,
+      final File targetFile) throws IOException {
+    try {
+      impersonator.doAs(artifactId, () -> {
+        Locations.linkOrCopy(artifactDetail.getDescriptor().getLocation(), targetFile);
+        return null;
+      });
+    } catch (FileAlreadyExistsException ex) {
+      LOG.warn("Artifact file {} already exists.", targetFile.getAbsolutePath());
+    } catch (Exception e) {
+      Throwables.propagateIfPossible(e, IOException.class);
+      // should not happen
+      throw Throwables.propagate(e);
+    }
+  }
+
+  private Map<String, String> getDistributedExtraProgramOptions() {
+    return Collections.emptyMap();
+  }
+
+  private Map<String, String> getExtraProgramOptions() {
+    return Collections.singletonMap(ProgramOptionConstants.HOST, hostname);
```

Comment:
InMemoryProgramRuntimeService returns `Collections.singletonMap(ProgramOptionConstants.HOST, hostname);` while AbstractProgramRuntimeService returns `Collections.emptyMap();`. 
Why always returning Collections.singletonMap? This is one of the reasons the tests are failing.

Reply:
Also, merge getDistributedExtraProgramOptions and getExtraProgramOptions